### DISCUSSION TITLE
2023-08-18 - Testing 3 - Automating Tests - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#6374] [Chewing the Fat]
Hey testers!

Let's talk about [this rule](https://www.ssw.com.au/rules/good-candidate-for-automation/)

**Comment by brydeno** [comment]
We do all this testing to reduce the risk of introducing unwanted behaviour. Can we quantify how much risk is reduced? Or is knowing our risk profile entirely a qualitative question.

**Comment by ozdemir-mehmet** [comment]
I didn't write many unit tests before CER, but since being "forced" to, I'm converted to writing unit tests to cover all CRM related plugins, api code etc.

**Comment by drwharris** [comment]
Automated tests are pointless if the author doesn't validate the outcomes. To many times I have seen developers copy and paste a similar tests, change a few things to match what they think is correct and then run the test. It passes and they move on. Not realising that they have a false positive. 

The other issue I see often is people write a single test to cover single positive path. Never testing alternate paths and conditions.

**Comment by leahy268** [comment]
The best I've got here is I've setup the pipeline to run a developers tests and if they pass then I let it deploy.  Whether the tests were useful or even did what they were meant to do is something that I can't answer.  I haven't as yet written a test :).

**Comment by leehawkinsssw** [comment]
The rule gives some good high-level tips about making decisions about where automation might be helpful. 

But, remember, it's not automated **or** human testing, it's automated **and** human testing - working out the appropriate mix of these is the key to good test strategy. It should always be a decision based around risks and the best approach to mitigating those risks.

**Comment by jakebayliss** [comment]
I am not sure I agree with "Tests of early stage features" being in the do not automate category. Seems to violate TDD. I would create tests if it is an early stage feature or not. Not only does this validate the feature being built but is also a way to run the code without much effort. Most of the time it is not too difficult to deal with a changing feature

**Reply by tomek-i** [reply]
Most of the times that's also where 💩 hits the fan where you think "ahhh, that's an easy change" and suddenly all kinds of things break down the line and on other components that you did not expect to break.

**Reply by wicksipedia** [reply]
There's TDD and there's pragmatic TDD

**Comment by Khaled-Albahsh** [comment]
not really the biggest guru when it comes to testing. Whatever looks feasible and fruitful at the moment I need to test is what I use.

**Comment by divgul** [comment]
Excellent CTF and rule. Very informative and well structured.

**Comment by ncn-ssw** [comment]
Automated tests come into their own when coupled with a fully automated deployment pipeline.

**Comment by BenDunkerleySSW** [comment]
Although I'm not a developer, I am curious how the quality of automatic testing can be guaranteed. If it is through significant human scrutiny, then what is the actual resource difference between a fully manual test and an automatic test - that being, If an automated test requires a human to monitor/check its outcome every time it's run, how much time will actually be saved?

**Reply by GordonBeeming** [reply]
Howzit @BenDunkerleySSW, 
The overall outcome of each automated test (or check) is pass or fail, the individual test outcome would be checked in each test, and generally, as long as the tests have been peer reviewed there's a higher success rate that the correct outcomes are being checked for
With this in mind, tests are run in the products pipeline and if a test fails, the overall pipeline fails. 
So as long as no functionality is breaking a test, nobody is specifically checking if the tests are working because the pipeline is not breaking.
Note: Devs should be running tests locally as they go so the pipeline shouldn't break, to speed this cycle up, some development environments can run tests in the background while coding so you get early signals when tests are breaking as well

**Reply by leehawkinsssw** [reply]
> "If an automated test requires a human to monitor/check its outcome every time it's run"

This scenario suggests the tests are poor and the requirement for significant human intervention is an anti-pattern. The following rules might also be helpful:

https://www.ssw.com.au/rules/automated-test-code-first-class-citizen/
https://www.ssw.com.au/rules/dangers-of-tolerating-test-failures/
https://www.ssw.com.au/rules/review-automated-tests/

**Comment by joshbermanssw** [comment]
Even though i haven't come across testing for clients yet in FBC, both the video and the rule page were very easily digestible and informative. I think it does need to be emphasised though that test-cases will only be effective when written with accuracy, and I am curious about the process behind validating this.

**Comment by vladislav-kir** [comment]
I great rule! I really liked the examples at the end :)

**Comment by AlexRothwell** [comment]
The rule conveys a pragmatic approach to automated testing quite well. I think it'd be quite useful to show to clients if they're pushing for fully automated testing so you can find a more appropriate blend of testing

**Comment by danielmackay** [comment]
I liked this rule, particularly around the points of what SHOULD and SHOULD NOT have automated tests.

I also come across a post on LinkedIn recently on Unit Test best practices which is somewhat related:  https://www.linkedin.com/feed/update/urn:li:activity:7097819379819589632/?origin=SHARED_BY_YOUR_NETWORK

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/2636640/71ec2008-8ebd-4758-ad89-307e8e7f1d9a)

**Comment by jeoffreyfischer** [comment]
Great explanation about what automated tests are!

**Comment by Ben0189** [comment]
i have gain better understanding on automated testing and when to do automate test from the rules

**Comment by tomek-i** [comment]
https://osherove.com/ good resources for testing, TDD etc. has also lots of youtube videos

**Comment by jomelsontanco** [comment]
In my client projects outside of SSW, I find that political considerations not techincal ones drive automation. Client decision makers are reluctant to invest in automation that wouldn't produce an ROI short-term. I am interested in seeing a CTF that provides guidelines on how to present ROIs of automated testing to potential clients.

**Comment by chrisschultzssw** [comment]
As a SysAdmin, I'm not doing any developer level testing... though any tests I do are unlikely to be repeated, so I stick with manual tests.

**Comment by AntPolkanov** [comment]
Great rule. I especially enjoyed the "Further Reading" section! Was it always a part of the rules, or is it a recent addition? It keeps the rule concise while providing an opportunity to explore the topic.

**Comment by sethdaily** [comment]
Interesting to think about what should and shouldn't be automated (this has far-ranging applications nowadays). It sounds like it would be hard to reach a point where all tests are automated - there is always use cases for the human.

**Comment by amankumarrr** [comment]
Great explanation on automating testing based on different use cases.

**Comment by KahaMason** [comment]
Great Rule & good informative examples.

**Comment by Jord-Gui** [comment]
Good rule - never thought about the types of tests that you shouldn't automate

**Comment by Harry-Ross** [comment]
Thought it was a great rule - good to think about whether the investment of creating automated tests may not be worth it when compared to manual testing.

**Comment by JackDevAU** [comment]
Testing is all about time vs output. Is the test worth it, I think this rule gives some good guidelines to follow to figure out that question! 

✅Rule hits the main points, when to test and when not to test
✅YT video was quite good!

**Comment by zacharykeeping** [comment]
Great rule - very detailed and informative, and gives a good explanation for when and when not to do automated testing.

**Comment by Abf09** [comment]
After doing CTF this week, I learned when do we automate tests ad when we don't which is crucial when working with clients. It'd good see data of how often we automate tests for our internal projects.

**Comment by cvionnet** [comment]
Great rule, especially the good/don't/maybe section

**Comment by AshAnil** [comment]
Nice rule - It is nicely shown what can we do with Automating tests and what we can't do.

**Comment by AttackOnMorty** [comment]
Automated testing is not enough. It needs to be combined with [exploratory testing](https://en.wikipedia.org/wiki/Exploratory_testing) to see if it really covers all we need.

**Reply by leehawkinsssw** [reply]
You might like to check out this rule too: https://www.ssw.com.au/rules/what-is-exploratory-testing/

**Comment by ricksu978** [comment]
I like the graph in the video showing the cost of automated test vs manual test.

To get the best out of both world, "Balance" is the key.
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/35659961/83d27cf3-6077-4b2f-835d-e35f21d29d5c)

**Comment by piers-sinclair** [comment]
There is value in both automated and manual tests.

The key is to figure out whether the ROI is worth it and whether the functionality can be tested in a stable way.

**Comment by tombui99** [comment]
Awesome rule about automated testing. However, it is quite a debatable topic when it comes to whether or not is worth it to spend a lot of devs' time writing automated tests when requirements are often changing (especially if it's on the Front-end).

**Comment by jackreimers** [comment]
Very informative rule, it's good to have some documented examples of when and when not to setup automated tests vs manual tests.

**Comment by JeanThirion** [comment]
I didnt like the video - I think the balance graph (main point of that 2mn section) is both an oversimplification of the problem, and so obvious it shouldn't require a video. Please teach me something interesting !

Now automating tests is indeed important and tricky, and finding the balance is the challenge. I would have loved the video to give me tips about how to find that balance, rather than just oversimplifying the problem by stating "there's a balance". Thanks captain obvious!

**Reply by JeanThirion** [reply]
Oh and btw the rule is excellent at doing just that ⭐⭐⭐

**Comment by Dhruv-0987** [comment]
Testing is extremely important but is always overlooked by developers, but I learnt through todays discussion is what factors to consider while deciding whether the test should be automated or manual (manual here is writing unit and integration tests). The extent of testing using automated test is limited and never enough, thats why finding the right balance between the time required to write those tests and the time put in by testers to test the same functionality is crucial.

**Reply by leehawkinsssw** [reply]
Just to be clear, we should be thinking about "human" and "automated" testing to distinguish the testing performed by people and that performed by machines. Typically we write unit and integration tests with the intent of them being run by machines, so these would come under the "automated testing" category, not "manual"

**Comment by PennyWalker** [comment]
I really enjoyed this video, he's a great presenter! It taught me a lot about what testing entails and when to use it! I've always wondered how devs can balance the need to test via the time cost!

**Comment by jimmy1992abc** [comment]
The rule is well explained. I am not a developer, but from my perspective, what if the error that is ignored by the automated testing then make it even more unlikely to be picked up? 

It can be even harder if the automated testing solved the issue, but not all of them.
(People will question the mechanism if it keeps saying 100/100 good, but people won't question the other 97/100 if the mechanism tells you there are 3 errors)

Ideally, there would be more than 2 systems testing using different logic, so the change of double failure is squared.

**Reply by leehawkinsssw** [reply]
Thanks, @JimmyChenSSW These rules are somewhat relevant to your comment:

https://www.ssw.com.au/rules/automated-test-code-first-class-citizen/
https://www.ssw.com.au/rules/dangers-of-tolerating-test-failures/
https://www.ssw.com.au/rules/review-automated-tests/

**Comment by Marxoz** [comment]
Good rule! I will recommend watching the entire video in the CTF. There are some good explanations of were what types of tests fit in. 

Automated testing is really helpful to make sure the core of a system is healthy. But like Andrew said, as long as the tests are well written. You need to test your test too. Both positive and negative testing, otherwise it is pointless.

Also, I think @leehawkinsssw  should grow a similar beard to the guy in the video 🧔‍♂️

**Reply by leehawkinsssw** [reply]
Just to be clear, my👍applies to all but the last part of @Marxoz's comment 😄

**Comment by SamBWagner** [comment]
I think tests are a valuable tool which are used to add rigidity/security to the solution. That said, it's always best to take into account that tests are an optional tool which need only be added when that rigidity is valuable. 

If you're working in a type-safe language, with mature libraries and the scope of the project is small. Even unit tests can be more of a hinderance than a help.

That said, assume you should test, then find the exception is a good rule of thumb

**Comment by jernejk** [comment]
I like how automated testing is introduced, shows the balance between manual and automated testing and even shows when to use manual over automated testing.

I would also love to see, how this will impact our Spec Review process and PBI/bug/issues effort estimation. We want the clients to let us do more testing, do "Done videos" and lots of other stuff but we also need to be mindful that each component adds additional time to the developer and directly impacts how much work they can complete in a Sprint which also impacts the cost of the project.

**Comment by stephanfakossw** [comment]
Although I'm not a dev, I'm very keen on testing. In the accounting world, solid checks are very important. We would typically automate all the checks (as they are typically very repetitive), but then we would also pass it through a high level "make sense" check to confirm the numbers are accurate.

**Comment by babakamyljanovssw** [comment]
Very informative rule! I got understanding about what tests can or can't be automated.

**Comment by bahjatalaadelssw** [comment]
We might not get to automating edit test pleases, but we can automate much of the process now to speed up the workflow.

**Comment by jimmidier** [comment]
In my opinion it's not the subjectivity that negates automated tests, but rather the complexity of the definition to correctness.
Even in the case of UI there is still a set of potential standards under mainstream aesthetics that defines correctness to be "saturation under 0.8", "4-point grid system" and the like. It's just too difficult to form unified judgements for tests that we treat it as being a subjective thing.
That said, since subjectivity is the main factor of the high complexity it can still represent the attribute but is not by itself the very reason for manual testing.

**Comment by Aibono1225** [comment]
Helpful video and rule. We all know automating tests can be beneficial, but as we've mentioned, not all tests should be automated. This uncertainty often causes us to overlook its usage.

**Comment by william-liebenberg** [comment]
Im not a tester! :) But I really like this rule. Especially useful to help new team members and client devs when it comes to deciding what to automate vs not.

**Comment by KristenHu** [comment]
The rule is really useful, I didn't write many unit test before, and it helps me to figure out whether I should write a test or not

**Comment by tino-liu** [comment]
Automated tests are so nice to have to keep the overall integrity of the functionality side within the application. Manual testing is also valuable to ensure the UI/UX side stays in shape and continue to provide users the best experience.

**Comment by jerryluossw** [comment]
Automating Tests is awsome, but for client project, it depends on whether client likes to pay for it.
If it is, we will add  Automating Tests for some core and important business logic.

By the way, One point that's easy to overlook is that sometimes  Automating Tests is better than document even comments in code, because document and comments are easy to be outdated, but if we have a good Automating Tests case, even 10 years later, new dev also can get the correct business logic from the Automating Tests.

**Comment by matt-goldman** [comment]
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/19944129/2079555e-e7b5-4ac5-9c1e-2a092f50afec)
Automated testing is awesome. But the question to automate comes a long way down the road from first deciding what to test. Once you know what needs to be tested, then you can figure out _how_.

**Reply by leehawkinsssw** [reply]
Amen to that, @matt-goldman 💯

**Comment by johnxu16** [comment]
A great rule to read before deciding which kind of testing we should use in the project and whether it needs to be implemented as an automatic test.

**Comment by Freego1783** [comment]
It's important to thing about what should or should not be automated and it's nice to have some insights about that !

**Comment by Hona** [comment]
A lot of people never stop to think of if they should, not how they can. I like these rules to add as ammunition to internal guys or clients when they are asking about testing or going down the wrong track

**Comment by steven0x51** [comment]
I think manual testing should be always required and the coverage of automated tests depends on the size and scalability of the project. Lager project should have more automated test coverage.

**Comment by BrookJeynes** [comment]
Automated and manual testing is important but it's important to weigh up the pros and cons about whether that feature really does need to be tested.

**Comment by tkapa** [comment]
When you start getting this deep into Unit testing it becomes clearer that it needs to be given enough room that it becomes almost entirely separate to developing the application.

**Comment by jontygardner** [comment]
It is hard to find areas in the TV Team where we can automate our testing. Videos are too subjective to put through an automation test. However, we implement spell checkers and Grammarly to 'test' our scripts, subtitle tracks, and graphics on our videos.

**Comment by Levijj22** [comment]
Great examples in the rule, but I didn’t feel engaged when watching the video. Many people believe that every process should be automated, however as mentioned, not everything should be automated, including some tests. It seems there will always be a need for human intervention. (which I’m sure Testers are happy to hear).

**Comment by TobyChurches** [comment]
Great Rule! Will provide a good reference point when deciding whether a change is worth testing

**Comment by MattParkerDev** [comment]
Automated tests = good
